src/primitives/hrp.rs

```diff
@@ -119,6 +119,78 @@ impl Hrp {
         Ok(new)
     }
 
+    /// Parses the human-readable part from an object which can be formatted.
+    ///
+    /// The formatted form of the object is subject to all the same rules as [`Self::parse`].
+    /// This method is semantically equivalent to `Hrp::parse(&data.to_string())` but avoids
+    /// allocating an intermediate string.
+    pub fn parse_display<T: core::fmt::Display>(data: T) -> Result<Self, Error> {
+        use Error::*;
+
+        struct ByteFormatter {
+            arr: [u8; MAX_HRP_LEN],
+            index: usize,
+            error: Option<Error>,
+        }
+
+        impl core::fmt::Write for ByteFormatter {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                let mut has_lower: bool = false;
+                let mut has_upper: bool = false;
+                for b in s.bytes() {
+                    // Continue after finding an error so that we report the first invalid
+                    // character, not the last.
+                    if !b.is_ascii() {
```

Comment:
There are a few problems here, I uncovered them by writing a test to check the code comment (which is incorrect, we return the _last_ error), I think we want `break` instead of `continue`.

We need to iterate over `chars` and check they are ascii instead of using `bytes` - we do this in `parse` above.

Reply:
Instead of commenting each line I just wrote some tests that I think should pass

```rust

    #[test]
    fn parse_non_ascii() {
        assert_eq!(Hrp::parse("❤").unwrap_err(), Error::NonAsciiChar('❤'));
    }

    #[test]
    fn parse_display_non_ascii() {
        assert_eq!(Hrp::parse("❤").unwrap_err(), Error::NonAsciiChar('❤'));
    }

    #[test]
    fn parse_display_returns_first_error() {
        assert_eq!(Hrp::parse_display("❤ ").unwrap_err(), Error::NonAsciiChar('❤'));
    }

    // This test shows that the error does not contain heart.
    #[test]
    fn parse_display_iterates_chars() {
        // FIXME: Once test above passes the error char will need changing.
        assert_eq!(Hrp::parse_display(" ❤").unwrap_err(), Error::NonAsciiChar('❤'));
    }
```

Reply:
Thanks! I will take these tests and make them pass.